File: copilot-bot/src/commands/history.ts
import type { CommandHandler } from "@woife5/shared";
import { type ChatInputCommandInteraction, PermissionFlagsBits, SlashCommandBuilder } from "discord.js";
import { clearHistory, getHistory } from "llm-connector/chat-history";
import { adminEmbed } from "./embed";

export const history: CommandHandler = {
    data: new SlashCommandBuilder()
        .setName("history")
        .setDescription("Edit the chat history.")
        .setDefaultMemberPermissions(PermissionFlagsBits.Administrator)
        .addStringOption((option) =>
            option
                .setName("command")
                .setDescription("The command to execute on the chat history.")
                .setRequired(true)
                .addChoices({ name: "clear", value: "clear" }, { name: "length", value: "length" }),
        ),
    executeInteraction: async (interaction: ChatInputCommandInteraction): Promise<void> => {
        const command = interaction.options.getString("command", true);

        switch (command) {
            case "length": {
                const embedLength = adminEmbed().setDescription(
                    `Current chat history length: \`${getHistory().length}\``,
                );
                interaction.reply({ embeds: [embedLength] });
                return;
            }
            case "clear": {
                clearHistory();
                const embed = adminEmbed().setDescription(`Chat history has been cleared.`);
                interaction.reply({ embeds: [embed] });
                return;
            }
        }
    },
};

File: copilot-bot/commands/history.go
package commands

import (
	"fmt"

	"github.com/bwmarrin/discordgo"
	"github.com/woife5/copilot-bot/config"
)

func (b *Bot) handleHistory(s *discordgo.Session, i *discordgo.InteractionCreate) {
	options := i.ApplicationCommandData().Options
	if len(options) == 0 {
		return
	}

	command := options[0].StringValue()

	switch command {
	case "length":
		embed := adminEmbed()
		embed.Description = fmt.Sprintf("Current chat history length: `%d`", b.History.Length())
		s.InteractionRespond(i.Interaction, &discordgo.InteractionResponse{
			Type: discordgo.InteractionResponseChannelMessageWithSource,
			Data: &discordgo.InteractionResponseData{
				Embeds: []*discordgo.MessageEmbed{embed},
			},
		})

	case "clear":
		b.History.Clear()
		embed := adminEmbed()
		embed.Description = "Chat history has been cleared."
		s.InteractionRespond(i.Interaction, &discordgo.InteractionResponse{
			Type: discordgo.InteractionResponseChannelMessageWithSource,
			Data: &discordgo.InteractionResponseData{
				Embeds: []*discordgo.MessageEmbed{embed},
			},
		})
	}
}

// adminEmbed returns a base embed for admin command responses.
func adminEmbed() *discordgo.MessageEmbed {
	return &discordgo.MessageEmbed{
		Color: 0x00FFFF, // Aqua
		Author: &discordgo.MessageEmbedAuthor{
			Name:    "Copilot",
			IconURL: config.CopilotIcon,
		},
	}
}
